src/api/auth.ts

```diff
@@ -0,0 +1,270 @@
+import Cookies, { CookieSetOptions } from "universal-cookie";
+import { AxiosResponse } from "axios";
+
+import APIClient from "./client";
+
+const OAUTH2_CLIENT_ID = process.env.REACT_APP_OAUTH2_CLIENT_ID;
+const PRODUCTION = process.env.NODE_ENV !== "development";
+
+/**
+ * Authorization result as returned from the backend.
+ */
+interface AuthResult {
+    username: string,
+    expiry: string
+}
+
+/**
+ * Name properties for authorization cookies.
+ */
+enum CookieNames {
+    Scopes = "DiscordOAuthScopes",
+    Username = "DiscordUsername"
+}
+
+export interface APIErrors {
+    Message: APIErrorMessages,
+    Error: any, /* eslint-disable-line @typescript-eslint/no-explicit-any */
+}
+
+export enum APIErrorMessages {
+    BackendValidation = "Backend could not authorize with Discord. Please contact the forms team.",
+    BackendValidationDev = "Backend could not authorize with Discord, possibly due to being on a preview branch. Please contact the forms team.",
+    BackendUnresponsive = "Unable to reach the backend, please retry, or contact the forms team.",
+    BadResponse = "The server returned a bad response, please contact the forms team.",
+    Unknown = "An unknown error occurred, please contact the forms team."
+}
+
+/**
+ * [Reference]{@link https://discord.com/developers/docs/topics/oauth2#shared-resources-oauth2-scopes}
+ *
+ * Commented out enums are locked behind whitelists.
+ */
+export enum OAuthScopes {
+    Bot = "bot",
+    Connections = "connections",
+    Email = "email",
+    Identify = "identify",
+    Guilds = "guilds",
+    GuildsJoin = "guilds.join",
+    GDMJoin = "gdm.join",
+    MessagesRead = "messages.read",
+    // RPC = "rpc",
+    // RPC_API = "rpc.api",
+    // RPCNotifRead = "rpc.notifications.read",
+    WebhookIncoming = "webhook.incoming",
+    // AppsBuildsUpload = "applications.builds.upload",
+    AppsBuildsRead = "applications.builds.read",
+    AppsStoreUpdate = "applications.store.update",
+    AppsEntitlements = "applications.entitlements",
+    // RelationshipsRead = "relationships.read",
+    // ActivitiesRead = "activities.read",
+    // ActivitiesWrite = "activities.write",
+    AppsCommands = "applications.commands",
+    AppsCommandsUpdate = "applications.commands.update"
+}
```

Comment:
Do we need to list all possible scopes? Maybe only the ones we will absolutely use.
```suggestion
export enum OAuthScopes {
    Connections = "connections",
    Email = "email",
    Identify = "identify",
    Guilds = "guilds"
}
```